File: src/main/java/actions/MouseEventManager.java
package actions;

import ux.DrawRateManager;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class MouseEventManager {
    private static MouseEventManager instance;

    private MouseEventManager() {}

    public static MouseEventManager getInstance() {
        if (instance == null) {
            instance = new MouseEventManager();
        }
        return instance;
    }

    private final List<MouseEventReceiver> mouseEventReceivers = new ArrayList<>();

    public void addReceiver(MouseEventReceiver receiver) {
        mouseEventReceivers.add(receiver);
    }

    public void addAll(Collection<? extends MouseEventReceiver> receivers) {
        mouseEventReceivers.addAll(receivers);
    }

    public void removeReceiver(MouseEventReceiver receiver) {
        mouseEventReceivers.remove(receiver);
    }

    private boolean mousePressedOverAny;
    private MouseEventReceiver pressedReceiver;

    public void onMousePressed(int mouseX, int mouseY) {
        mousePressedOverAny = false;
        pressedReceiver = null;

        for (MouseEventReceiver receiver : mouseEventReceivers) {
            if (!mousePressedOverAny && receiver.mousePressedOverMe()) {
                mousePressedOverAny = true;
                pressedReceiver = receiver;
            }

            receiver.onMousePressed();

        }
    }

    public void onMouseReleased() {
        if (pressedReceiver != null) {
            pressedReceiver.onMouseReleased();
            DrawRateManager.getInstance().drawImmediately();
            pressedReceiver = null;
        }
    }

    public boolean isMousePressedOverAnyReceiver() {
        return mousePressedOverAny;
    }

    public boolean isMouseDraggedOverReceiver() {
        return (pressedReceiver != null);
    }

}

File: src/main/java/actions/MouseEventManager.kt
package actions

import ux.DrawRateManager

class MouseEventManager private constructor() {
    private val mouseEventReceivers: MutableList<MouseEventReceiver> = ArrayList()
    fun addReceiver(receiver: MouseEventReceiver) {
        mouseEventReceivers.add(receiver)
    }

    fun addAll(receivers: Collection<MouseEventReceiver>?) {
        mouseEventReceivers.addAll(receivers!!)
    }

    var isMousePressedOverAnyReceiver = false
        // kotlin thing - says that isMousePressedOverAnyReceiver can only be read externally but can be written to interanlly
        private set

    private var pressedReceiver: MouseEventReceiver? = null

    fun onMousePressed(mouseX: Int, mouseY: Int) {
        isMousePressedOverAnyReceiver = false
        pressedReceiver = null
        for (receiver in mouseEventReceivers) {
            if (!isMousePressedOverAnyReceiver && receiver.mousePressedOverMe()) {
                isMousePressedOverAnyReceiver = true
                pressedReceiver = receiver
            }
            receiver.onMousePressed()
        }
    }

    fun onMouseReleased() {
        if (pressedReceiver != null) {
            pressedReceiver!!.onMouseReleased()
            DrawRateManager.getInstance().drawImmediately()
            pressedReceiver = null
        }
    }

    companion object {
        @JvmStatic
        var instance: MouseEventManager? = null
            get() {
                if (field == null) {
                    field = MouseEventManager()
                }
                return field
            }
            private set
    }
}
